### ISSUE TITLE
Issues of expression and arithmetic operators

### ISSUE THREAD

**Issue by MollyDream** [verified-network-toolchain/petr4#261]
During the development of expressions ([Semantics.v](https://github.com/cornell-netlab/petr4/blob/poulet4/deps/poulet4/lib/Semantics.v)) and arithmetic operations ([Ops.v](https://github.com/cornell-netlab/petr4/blob/poulet4/deps/poulet4/lib/Ops.v) and [P4Arith.v](https://github.com/cornell-netlab/petr4/blob/poulet4/deps/poulet4/lib/P4Arith.v)) in P4Light, I found a series of discrepancies and issues in the Petr4 (checker.ml and ops.ml), P4Cub (BigStep.v and P4Arith.v), bmv2 model and P4 manual. This report contains issues in not just the arithmetic operations but other related codebases (like typechecker and expression semantics). I hope this can help achieve the final goal of sharing a common codebase of the arithmetic operations among Petr4, P4Light and P4Cub.
The manual version I am referring to is [P4-16-v1.2.1](https://p4.org/p4-spec/docs/P4-16-v1.2.1.html).

#### Concatenation
```
bit<8> x = 6w1 ++ 2w1;
```
1. The manual says that "concatenation is applied to two bit-strings (signed or unsigned)... and the sign of the result is taken from the left operand." (reference: 8.5, 8.6 and 8.6.1) However, in Petr4's typechecker (check.ml), all the four possible concatenations output the bit type, i.e. `int<> ++ int<> = bit <>`, `int<> ++ bit<> = bit<>`, `bit<> ++ int<> = bit<>` and `bit<> ++ bit<> = bit<>`, out of which the first two should have output `int<>` instead. Also, out of the four concatenations, only `bit<> ++ bit<>` is implemented in the Petr4 evaluator (ops.ml). For example, `6s1 ++ 2s1` is accepted by the checker, but the output type is bit<8> instead of int<8>, and it cannot be evaluated.
2. "Concatenation is applied to two bit-strings (signed or unsigned)." Also, among all the binary operators, shifts and concatenation are NOT listed among "the binary operations [that] require both operands to have the same exact type (signedness) and width". (reference: 8.5, 8.6 and 8.6.1) This implies that concatenation of fixed-width int/bit with arbitrary-precision integer is not allowed (i.e. `int<> ++ integer`, `bit<> ++ integer`, `integer ++ int<>` and `integer ++ bit<>` are illegal). However, Petr4 checker erroneously implicitly casts integer in these cases. For example, `bit<8> x = 4w5 + 1` is accepted by the checker with an implicit cast on the integer 1. 
**A more important message from the fact the operands of shift and concatenation do not need to have the same type and width is that no implicit casts should not be added to integers involved in shift and concatenation.**

#### Shift
```
bit<8> x = 8w1 << 1; 
```
1. The left operand of bit shift can be fixed-width bit, fixed-width int and arbitrary-precision integer (reference: 8.5, 8.6, 8.7), but the Petr4 typechecker (check.ml) does not allow arbitrary-precision integer as the left operand. For example, `10 >> 1` is rejected by Petr4.
2. The right operand of bit shift should be "either an unsigned number of type `bit<S>` or a non-negative integer literal", where a "non-negative integer literal" can be a non-negative arbitrary-precision integer (reference: 8.5, 8.6, bmv2). Petr4's typechecker (checker.ml) correctly allows only fixed-width bit and arbitrary-precision integer, but its evaluator implements erroneously fixed-width int in addition (ops.ml).
3. For the special case of `integer >> bit<>` (or `<<`), it is unclear whether or not it should be allowed. The description of integer's shift operations says that "the right operand must be constant and positive." The description about all the operations on arbitrary-precision integer says that "each operand that participates in any of these operation must have type int", which, unlike bit<> and int<>, does not exclude bit shift as a special case (reference: 8.5, 8.6, 8.7). Also, `integer >> bit<>` is listed as an example of illegal arithmetic expressions in 8.9.3 (`1 << ((bit<8>) x) ` is illegal). However, this operation is allowed in bmv2. Moreover, the description of "each operand that participates in any of these operation must have type int" may be legacy content, which also erroneously appears in 8.6 in a slightly paraphrased way ("all binary operations require both operands to have the exact same type"). Lastly, I don't see any reason it should not work. For example, the following example works in bmv2.
```
const bit<3> x = 1;
bit<8> y = 6 >> 1;
```

#### Bitwise operations
```
int<8> x = ~ 8s1;
```
1. The manual says that "bitwise operations on signed integers are well-defined, since the representation is mandated to be two's complement" (reference: 8.6). However, the bitwise operations (complement, add, xor, or) on fixed-width int are not allowed in Petr4's typechecker and not implemented in Petr4's evaluator. For example, `~ 8s1` and `8s1 | 8s0` are rejected by the Petr4 checker.

#### Seriablizble enum
```
enum bit<8> E {
   e1 = 0, e2 = 1, e3 = 2
}
```
1. The manual says "[serializable enum] requires the programmer provide both the fixed-width unsigned (or signed) integer type and an associated integer value for each symbolic entry in the enumeration." (reference: 7.2.1) This means that both fixed-width bit and int can serve as the underlying type, but neither of bmv2 and Petr4 accepts int<> type enum. For example, `enum int<8> E1 {e1 = 0}` raises a syntax error in Petr4.
2. The manual says that "because it is always safe to cast from an enum to its underlying fixed-width integer type, implicit casting from an enum to its fixed-width (signed or unsigned) integer type is also supported" and "implicit casting from an underlying fixed-width type to an enum is not supported." This means that whenever an enum (with an underlying type) appears in any operations, it can be flexibly cast to its underlying type. For example, 
```
enum bit<8> E1 {
   e1 = 0, e2 = 1, e3 = 2
}
enum bit<8> E2 {
   e1 = 10, e2 = 11, e3 = 12
}
E1 a = (E1) (E2.e1 + E2.e2);
bool b = (E2.e1 > E2.e2);
bit<8> = a << 3;
```
are all legal examples listed in the manual (reference: 8.3). However, Petr4's typechecker does not allow any of these. I think the key problem is some misunderstanding of the senum's implicit cast when implementing `cast_expression` in checker.ml. 
Note that the implicit cast of senum is subtly different from the implicit cast of arbitrary-precision integer in that
(1) The cast type of senum is always the underlying type of senum while the cast type of integer depends on the other operand in a binary operation (reference: 8.9.2);
(2) While implicit casts should not be applied to integer in shift and concatenation operations (please refer to the "concatenation section" before), implicit casts may be applied to senum involved in these operations (not explicitly said in the manual, but based on the above example `bit<8> = a << 3` taken from the manual).
3. It is not clearly defined in the manual, but if senum can be flexibly cast to its underlying type anytime, then all the operations allowed on the underlying type should be eligible on senum. For example, unary operations and bit slicing should be appliable to bit type senum, which are currently illegal in the Petr4 typechecker.

#### Duplicate names
1. Duplicate declarations with the same name of many language constructs, like enum, struct, header, headerunion, action and control, are allowed in the Petr4 typechecker. The list may not be complete. For example, the following code snippet is accepted by Petr4.
```
enum bit<8> E {
  e1 = 6
}
enum bit<8> E {
   one = 1
}
```
2. Duplicate fields with the same name of struct, header and headerunion, are allowed in the Petr4 typechecker. For example, the following code snippet is accepted by Petr4.
```
struct S {
    bit<32> a;
    bit<32> a;
}
header hdr {
    bit<8> g;
    bit<8> g;
}
header_union hu {
    hdr h;
    hdr h;
}
```
It is unclear why the Petr4 typechecker does not handle name duplication right now. Should we expect that in a future version, maybe?

#### Null
1. TypVoid is not allowed to have equality and inequality operations in the Petr4 typechecker (reference: type_has_equality_tests in checker.ml), but these operations are implemented in the evaluator on VNull(ops.ml). 

#### Record and struct
```
{ e1 = 0, e2 = 1, e3 = 2}
```
1. To differentiate the record and the struct in values (ValBaseRecord & ValBaseStruct) and types (TypRecord & TypStruct),  the former is a mapping from field names to values, while the latter is a typed entity with well-specified underlying field types. The syntax specifies that the former should be prepended with the name of a struct (or a header) unless it can be inferred from the context (reference: 8.12, 8.14, 8.15). With this background, it is unclear whether or not the record type can be directly compared for equality (==) or inequality (!=). The manual says that "structure-valued expressions can be used ... in comparisons ..." (reference: 8.12), where the structure-valued expression means the record in Petr4, but the examples given in the manual have all "cast" the record to the struct either explicitly (prepended with the name of a struct ) or explicitly (inferred from the other argument) before the equality comparison. In Petr4, TypRecord is allowed to have equality and inequality operations in the typechecker (reference: type_has_equality_tests in checker.ml), but these operations are not implemented in the evaluator on VRecord (ops.ml). For example, `{a = 1, b =2} == {b = 2, a = 1}` is accepted by the Petr4 checker, but it cannot be evaluated. BMV2 compiler does not accept this example. If the record should not be compared for equality, the Petr4 typechecker should reject that.

#### List and tuple
```
{ 0, false, 8w1}
```
1. Very similar to the record and struct in the previous section, to differentiate the list and the tuple,  the former is a numeric list of values, while the latter is a typed entity with well-specified underlying element types. The P4 manual does not specify anything about whether or not the list or the tuple can be compared for equality (==) or inequality (!=). In Petr4 and BMV2, both the list and the tuple are allowed to have equality and inequality operations in the typechecker, with well-defined evaluation. Nothing needs to be done until the manual is clarified.

#### Equality and inequality
```
bool x = 8w1 == 8w1;
```
1. There are a number of cases in the equality operation of Petr4 evaluator (ops.ml) that should output invalid other than false. For example, equality operation on two bit<> with different width is invalid instead of false. Similarly, equality operation on two structs with different field names is invalid instead of false. Even though the typechecker (checker.ml) checks the type equality, it would still be nice to follow the same constraints in the evaluator.

#### Bit slicing
```
bit<4> x = 8w16[3:0];
```
1. The left operand of bit slicing can be fixed-width bit and int based on the manual (reference: 8.5, 8.6). Bit slicing on an arbitrary-precision integer is not defined in the manual nor Petr4, but implemented in bmv2. In the above example, `bit<4> x = 16[3:0]` works in bmv2.

#### Header stack access
```
Mpls_h[5] mpls;
bool x = mpls[4].isValid();
```
1. The manual says that "a P4 compiler must give an error if an index value that is a compile-time constant is out of range" (reference: 8.17) However, the Petr4 typechecker does not check that, allowing index out of bound. In the above example, `bool x = mpls[5].isValid()` passes the typechecker.

#### Type casting
1. `int<S> <-> bit<W> (S != W)` The manual does not list this kind of casting as a valid operation, and it emphasizes that P4 "cannot change both sign and width" in casting with a counterexample (Reference: 8.9.1, 8.9.3). However, the Petr4 typechecker erroneously accepts this kind of casting in `cast_ok`, and the evaluator also works. 
2. `Senum<U> <-> Senum<U>` The manual explains this kind of casting as first implicitly casting the senum to an underlying fixed-width type and then explicitly casting the fixed-width type to the other senum (Reference: 8.3). Also, explicit casts of senum are allowed only on "casts between an enum with an explicit type and its underlying type" (Reference: 8.9.1). However, the Petr4 typechecker directly casts senum to senum erroneously. Similar to other problems of senum described in the section before, this is still caused by the misunderstanding of the senum's implicit cast when implementing `cast_expression` in checker.ml.
3. `U -> Senum<U>` This specific direction of casting is allowed in the Petr4's typechecker, but not implemented in the evaluator (ops.ml).
4. `Record -> Struct` and `List -> Struct` The P4 manual syntax specifies 
```
expression ...
    | '{' kvList '}'
    | '(' typeRef ')' expression
    ;
```
in the section of record (Reference: 8.12). "For a structure-valued expression [i.e. record] typeRef is the name of a struct or header type. The typeRef can be omitted if it can be inferred from context, e.g., when initializing a variable with a struct type." The Petr4's typechecker reflects this by allowing both `'{' kvList '}'` and `(struct) '{' kvList '}'` to be cast to the struct when possible. For example, the first two lines of the example below are allowed. However, it is unclear from the manual if the `expression` in `'(' typeRef ')' expression` can be the other possible initializer of the struct, i.e. the list type. The Petr4's typechecker and BMV2 does not allow putting a `typeRef` before a list initializer, so `S z =  (S) {6}` the example below is not allowed. However, I don't see any reason it should not work; in fact, Petr4 realizes the list initializer (`S z =  {6}`) by inserting the explicit cast in the type checker, leading to essentially `S z =  (S) {6}`. It may be appreciated to clarify whether or not `(struct) '{' expressionList '}'` is allowed in the P4 manual. To illustrate the difference this brings in, please look at the three comparisons at the end of the example.
```
struct S {
    bit<4> a;
}

S z = {a=6};            // Valid
S z = (S) {a=6};        // Valid
S z = {6};              // Valid
S z =  (S) {6};         // Invalid

bool b = (S) {a=0xF6} == {a=0x26};    // Valid comparison of the equality of two structs; b = true
bool b = (S) {6} == {6};        // Invalid in Petr4 and BMV2
bool b = {0xF6} == {0x26};            // Valid comparison of the equality of two lists; b = false
```
5. `Record -> Struct` and `List -> Tuple` The manual allows the struct/header/tuple to be nested in the struct and tuple, so the type casting should be recursive in this case. However, Petr4's evaluator (ops.ml) does not apply casting recursively on the components of struct and tuple.

### Problems in P4Cub:
I have expanded the original descriptions and added a bunch of pointers to the previous sections and Ops.v. Given that Rudy wants to just replace P4Cub/P4Arith.v with P4Light/P4Arith.v, I have also added some advice on how to add/change the relations in BigStep.v to fix the remaining issues. Also, casting integer to int in the checker may not be a good idea, and I talk about the reasons in the following descriptions.

1. There should be negation (-) and bitwise complement (\~) on both the bit type and the int type, with subtly different meanings (reference: 8.5, 8.6), but P4Cub semantics (BigStep.v) implements only the negation on the int type and the complement on the bit type. For more details, please look at the section "**Bitwise operations**" above. Please refer to `eval_unary_op` in **Ops.v** for the implementations.

2. Many bounds in P4Cub/P4Arith.v are incorrect. (This can be fixed after replacing P4Cub/P4Arith.v with P4Light/P4Arith.v)

3. Concatenation of `int<> ++ int<>` (`IntArith.bit_concat`) is incorrect in P4Cub/P4Arith.v. In addition, only `bit<> ++ bit<>` is implemented in BigStep.v, and all the other three possible concatenations, `int<> ++ int<> = int<>`, `int<> ++ bit<> = int<>` and `bit<> ++ int<> = bit<>`,  are not implemented. For more details, please read the "**Concatenation**" section above.  Please refer to `eval_binary_op_plusplus` in **Ops.v** for the implementations.

4. Bit shifts of `int<> >> int <>` and `bit<> >> bit <>` (and `<<`) are implemented in P4Cub, which are incomplete and incorrect. The left operands of shifts should be an integer, int<> or bit<>, while the right operands of shifts should be only a non-negative integer or bit<>. For more details, please read the "**Shift**" section above. You may want to consider making shifts some separate relation(s) in `expr_big_step`, similar to the relation of concatenation (`plusplus`). Please refer to `eval_binary_op_shift` in **Ops.v** for the implementations.

5. Arbitrary-precision integer does not exist in the P4Cub's value type, but casting integer to int in the typechecker is probably not a good idea since the operations allowed on arbitrary-precision integer and fixed-width int are actually very different. For example, bitwise operations (bitwise-and/or/xor), saturating addition/subtraction and concatenation should be allowed for fixed-width int but **NOT** allowed for arbitrary-precision integer, while division and modulo should be allowed for arbitrary-precision integer but **NOT** allowed for fixed-width int. Please refer to `eval_binary_op_integer` and `eval_binary_op_int`  in **Ops.v** for a complete list of differences.

6. The multiplication operation is currently not implemented in P4Cub, and you may want to add that operator to the AST.v. Please refer to `eval_binary_op_bit`, `eval_binary_op_int` and  `eval_binary_op_integer` in **Ops.v** for the implementations.

7. Many cases in comparing the equality of two values result in false even when the equality operation is invalid in P4Cub (`eqbv` in Value.v).  For example, equality operation on two bit<> with different width should be invalid instead of false; similarly, equality operation on two records/headers with different field names should be invalid instead of false. Even though Check.v may check the type equality, it would still be nice to follow the same constraints in the semantics. For more details, please read the "**Equality and inequality**" section above. Please look at `eval_binary_op_eq` in **Ops.v** for the implementations.

8. Enum and senum do not exist in the P4Cub's value type, but this may be rationalized since no (s)enum declarations are in the AST.v.

9. Bit slicing is currently not implemented in the P4Cub's semantics, and you may want to add that operation. For more details, please read the "**Bit slicing**" section above.

10. In type casting of `int<S> <-> bit<W>`, `eval_cast` has output even when the casting operation is invalid in P4Cub (BigStep.v). The manual emphasizes that P4 "cannot change both sign and width" in casting with a counterexample (Reference: 8.9.1, 8.9.3). Therefore, when `S != W`, the casting should be invalid, but P4Cub does not reject that in `eval_cast`. Even though it is possible to check this in the checker (I didn't look at the Check.v to confirm if it has been implemented), it would still be nice to follow the same constraints in the semantics. For more details, please read the "**Type casting**" section above. Please look at `eval_cast` in **Ops.v** for the implementations.

11. The type casting from a tuple to a record/header is currently not implemented in P4Cub even though the manual allows initializing a struct and a header with a tuple (more accurately a list of values) (Reference: 8.11, 8.16). For more details, please read the "**Type casting**" section above. Please look at `eval_cast` in **Ops.v** for the implementations.

12. There are probably other issues that I didn't look into since the P4Cub type checker cannot be run right now.
 

### Problems in P4 manual:
1. Division (/) and modulo (%) operations on bit types are allowed and defined in Petr4 and bmv2, but not in the list of operations on fixed-width bit in 8.5.
2. The indices of bit slicing and header stack access are not restricted to any type in 8.5, 8.6 and 8.17, which may be fixed-width bit, fixed-width int, and arbitrary-precision integer based on the best guess from Petr4.
3. It would be helpful if the special cases where implicit casts of integers are illegal be explicitly stated in _8.9.2 implicit casts_. That section says that "applying a binary operation to an expression of type int and an expression with a fixed-width type will implicitly cast the int expression to the type of the other expression", but bit shifts are excluded (mentioned in 8.5 and 8.6), and concatenation are also excluded (8.6.1 does not mention integer as a valid operand of concatenation). Also, the following paragraph appearing in 8.5 and slightly paraphrased in 8.6 should probably be edited to include concatenation as a special case.
_All binary operations (except shifts) require both operands to have the same exact type and width; supplying operands with different widths produces an error at compile time. No implicit casts are inserted by the compiler to equalize the widths. There are no binary operations that combine signed and unsigned values (except shifts)._
4. In _8.3. operations on enum types_, the manual says that "implicit casting from an enum to its fixed-width (signed or unsigned) integer type" is supported, but it is unclear when exactly that is doable. An example in that sections shows that implicit casts of enum in bit shift are legal, what about concatenation, unary operations and bit slicing? Since enum is different from arbitrary-width integer in that it carries a deterministic underlying type, are all the operations allowed on the underlying type also eligible on enum?
5. For the special case of `integer >> bit<>` (or `<<`), it is unclear whether or not it should be allowed. The description of integer's shift operations says that "the right operand must be constant and positive." The description about all the operations on arbitrary-precision integer says that "each operand that participates in any of these operation must have type int", which, unlike bit<> and int<>, does not exclude bit shift as a special case (reference: 8.5, 8.6, 8.7). Also, `integer >> bit<>` is listed as an example of illegal arithmetic expressions in 8.9.3 (`1 << ((bit<8>) x) ` is illegal). However, this operation is allowed in bmv2. Moreover, the description of "each operand that participates in any of these operation must have type int" may be legacy content, which also erroneously appears in 8.6 in a slightly paraphrased way ("all binary operations require both operands to have the exact same type"). Lastly, I don't see any reason it should not work. For example, the following example works in bmv2.
6. To differentiate the record (i.e. structure-valued expression) and the struct, the former is a mapping from field names to values, while the latter is a typed entity with well-specified underlying field types. The syntax specifies that the record should be prepended with the name of a struct (or a header) unless it can be inferred from the context (reference: 8.12, 8.14, 8.15). With this background, it is unclear whether or not the record type can be directly compared for equality (==) or inequality (!=). The manual says that "structure-valued expressions can be used ... in comparisons ..." (reference: 8.12), but the examples given in the manual have all "cast" the record to the struct either explicitly (prepended with the name of a struct) or explicitly (inferred from the other argument) before the equality comparison. It should be clarified that whether or not two records, without any clue about the struct/header names, can be compared directly for equality and inequality. For example, is `{a = 8w1, b =2} == {b = 2, a = 8w1}` valid?
7. Very similar to the record and struct in the previous section, to differentiate the list and the tuple,  the former is a numeric list of values, while the latter is a typed entity with well-specified underlying element types  (reference: 8.10, 8.11). The P4 manual does not specify anything about whether or not the list or the tuple can be compared for equality (==) or inequality (!=). Currently, in Petr4 and BMV2, both the list and the tuple have equality and inequality operations.
8. The P4 manual syntax specifies
```
expression ...
    | '{' kvList '}'
    | '(' typeRef ')' expression
    ;
```
in _8.12. Structure-valued expressions_. "For a structure-valued expression [i.e. record] typeRef is the name of a struct or header type. The typeRef can be omitted if it can be inferred from context, e.g., when initializing a variable with a struct type." However, it is unclear from the manual if the `expression` in `'(' typeRef ')' expression` can be the other possible initializer of the struct, i.e. the list type (Reference: 8.11). However, I don't see any reason it should not work; in fact, Petr4 realizes the list initializer (`S z =  {6}`) by inserting the explicit cast in the type checker, leading to essentially `S z =  (S) {6}`.  It may be appreciated to clarify whether or not `(struct) '{' expressionList '}'` is allowed in the P4 manual. To illustrate the impact of this feature, please look at the following example.
```
struct S {
    bit<4> a;
}

S z = {a=6};            // Valid
S z = (S) {a=6};        // Valid
S z = {6};              // Valid
S z =  (S) {6};         // Invalid in Petr4 and BMV2

bool b = (S) {a=0xF6} == {a=0x26};    // Valid comparison of the equality of two structs; b = true
bool b = (S) {6} == {6};        // Invalid in Petr4 and BMV2
bool b = {0xF6} == {0x26};            // Valid comparison of the equality of two lists; b = false
```
9. In _8.9. Casts_, the manual says that "casts are only permitted between base types." This is apparently not true, contradicted by the three following casts given in _8.9.1. Explicit casts_, where the new types and senum are all not base types.
- casts between two types that are introduced by typedef and are equivalent to one of the above combinations
- casts between a type introduced by type and the original type.
- casts between an enum with an explicit type and its underlying type.
In addition, given the background information in bullet point 8, it is unclear whether prepending the structure-valued expression with typeRef is a kind of explicit cast, and whether omitting typeRef when inferrable is a kind of implicit cast. It would be nice to formalize this and clarify the situations where the explicit and implicit casts are allowed on struct and header type.

**Comment by rudynicolop** [comment]
`p4cub` will adopt the arithmetic operations in `p4light`'s `P4Arith.v` on `Z`.

`p4cub` does have reflected decidable and propositional equality for both values and expressions! Equality for expressions is modulo `tags_t`, whereas `p4cub` values have no `tags_t` so syntactic equality is sufficient.

**Comment by MollyDream** [comment]
@rudynicolop To help you better understand the issues, I have expanded the previous descriptions and added a bunch of pointers to the previous sections and Ops.v. Given that you want to just replace P4Cub/P4Arith.v with P4Light/P4Arith.v, I have also added some advice on how to add/change the relations in BigStep.v to fix the remaining issues. Also, casting integer to int may not be a good idea, and I talk about the reasons in the following descriptions.

1. There should be negation (-) and bitwise complement (\~) on both the bit type and the int type, with subtly different meanings (reference: 8.5, 8.6), but P4Cub semantics (BigStep.v) implements only the negation on the int type and the complement on the bit type. For more details, please look at the section "**Bitwise operations**" above. Please refer to `eval_unary_op` in **Ops.v** for the implementations.

2. Many bounds in P4Cub/P4Arith.v are incorrect. (This can be fixed after replacing P4Cub/P4Arith.v with P4Light/P4Arith.v)

3. Concatenation of `int<> ++ int<>` (`IntArith.bit_concat`) is incorrect in P4Cub/P4Arith.v. In addition, only `bit<> ++ bit<>` is implemented in BigStep.v, and all the other three possible concatenations, `int<> ++ int<> = int<>`, `int<> ++ bit<> = int<>` and `bit<> ++ int<> = bit<>`,  are not implemented. For more details, please read the "**Concatenation**" section above.  Please refer to `eval_binary_op_plusplus` in **Ops.v** for the implementations.

4. Bit shifts of `int<> >> int <>` and `bit<> >> bit <>` (and `<<`) are implemented in P4Cub, which are incomplete and incorrect. The left operands of shifts should be an integer, int<> or bit<>, while the right operands of shifts should be only non-negative integer or bit<>. For more details, please read the "**Shift**" section above. You may want to consider making shifts some separate relation(s) in `expr_big_step`, similar to the relation of concatenation (`plusplus`). Please refer to `eval_binary_op_shift` in **Ops.v** for the implementations.

5. Arbitrary-precision integer does not exist in the P4Cub's value type, but casting integer to int in the typechecker is probably not a good idea since the operations allowed on arbitrary-precision integer and fixed-width int are actually very different. For example, bitwise operations (bitwise-and/or/xor), saturating addition/subtraction and concatenation should be allowed for fixed-width int but **NOT** allowed for arbitrary-precision integer, while division and modulo should be allowed for arbitrary-precision integer but **NOT** allowed for fixed-width int. Please refer to `eval_binary_op_integer` and `eval_binary_op_int`  in **Ops.v** for a complet list of differences.

6. The multiplication opeartion is not implemented, and you may want to add that to the SAT.v. Please refer to `eval_binary_op_bit`, `eval_binary_op_int` and  `eval_binary_op_integer` in **Ops.v** for the implementations.

7. Many cases in comparing the equality of two values results in false even when the equality operation is invalid (`eqbv` in Value.v).  For example, equality operation on two bit<> with different width should be invalid instead of false; similarly, equality operation on two Record/Header with different field names should be invalid instead of false. Even though the typechecker may checks the type equality, it would still be nice to follow the same constraints in the semantics. For more details, please read the "**Equality and inequality**" section above. Please look at `eval_binary_op_eq` for the implementations.

8. Enum and senum do not exist in the P4Cub's value type, but this may be rationalized since no (s)enum declarations are ever allowed in the AST.v.

9. Bit slicing is not allowed in the P4Cub's semantics. For more details, please read the "**Bit slicing**" section above.

10. There are probably other issues that I didn't look into since the P4Cub type checker cannot be run right now.

**Comment by rudynicolop** [comment]
## P4cub Updates!

### P4cub uses P4light's `lib/P4Arith.v`
- `p4cub` uses `Z` in all places `N` previously occupied.
- `lib/P4cub/P4Arith.v` no longer exists.
- Evaluation of `p4cub` unary & binary operations for fixed-width integers (`bit<w>` & `int<w>`) use implementations from `lib/P4Arith.v`.
- `p4cub`'s static & dynamic semantics has a more comprehensive case-analysis for unary & binary operations, lifting the aforementioned restrictions on `~`, `_`, `++`, `<<`, `>>`.
- `p4cub`'s case-analysis for unary & binary operations is analogous to that of `lib/Ops.v`.
- Multiplication & bit-slicing now exist in `p4cub`.
- Valid types for unary operations are specified in `uop_type` in `P4cub/Check.v`
- Valid types for binary operations are specified in `bop_type` in `P4cub/Check.v`

### P4cub Casting
- `p4cub` now supports the casts `tuple -> header` & `tuple -> record`.
- All `p4cub` casts still must be explicit.
- Valid types for casting from `t1` to `t2` are specified by `proper_cast t1 t2` in `P4Cub/Check.v`

**Comment by hackedy** [comment]
I tried to turn everything into tests on the issue-261 branch in git. Anything
that doesn't fit in a test I'm noting here with a checkbox.

- [ ] The evaluator allows `e2` in `e1 << e2` to be a fixed-width int and it shouldn't.
- [ ] The integer >> bit<> stuff is listed as an example of illegal expressions. We should open a bug with p4c if it's not allowed. I think I may have already done this... however I added a shift-int-by-bit.p4 test anyway and we can remove it if we think it shouldn't be allowed.
- [ ] The evaluator shouldn't implement operations for VNull.
- [ ] We need to open a bug at the spec about records and lists.
- [ ] The evaluator does equality checks for things like 8w1 == 7w1 and doesn't need to. This means inserting runtime checks for things the type system enforces, but it should be really simple to do it so it's fine.
- [ ] We should open an issue about bmv2 implementing slicing for `int` values.

**Comment by hackedy** [comment]
OK, I've got all the tests fixed on the issue-261 branch and I'll try to get to the checkbox issues tomorrow.

**Comment by QinshiWang** [comment]
I'm cleaning up `admit`s in the repos. Can we remove `Mod` and `Div` from `Ops.eval_binary_op_bit`, so unsigned fixed-width integers cannot use `Div` and `Mod`?
https://github.com/verified-network-toolchain/petr4/blob/poulet4/deps/poulet4/lib/P4light/Semantics/Ops.v#L104

**Comment by hackedy** [comment]
That's OK with me.

**Comment by QinshiWang** [comment]
> I'm cleaning up `admit`s in the repos. Can we remove `Mod` and `Div` from `Ops.eval_binary_op_bit`, so unsigned fixed-width integers cannot use `Div` and `Mod`? https://github.com/verified-network-toolchain/petr4/blob/poulet4/deps/poulet4/lib/P4light/Semantics/Ops.v#L104

This is done.

**Comment by hackedy** [comment]
Closing. If any of these sub-issues crop back up please open a separate issue for them.
